pkg/reconcilers/base_reconciler.go

```diff
@@ -0,0 +1,113 @@
+package reconcilers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/3scale/apicast-operator/pkg/k8sutils"
+)
+
+// MutateFn is a function which mutates the existing object into it's desired state.
+type MutateFn func(existing, desired k8sutils.KubernetesObject) (bool, error)
+
+func CreateOnlyMutator(existing, desired k8sutils.KubernetesObject) (bool, error) {
+	return false, nil
+}
+
+type BaseReconciler struct {
+	// client should be a split client that reads objects from
+	// the cache and writes to the Kubernetes APIServer
+	client client.Client
+	// apiClientReader should be a client that directly reads objects
+	// from the Kubernetes APIServer
+	apiClientReader client.Reader
+	scheme          *runtime.Scheme
+	logger          logr.Logger
+}
+
+func NewBaseReconciler(client client.Client, apiClientReader client.Reader, scheme *runtime.Scheme, logger logr.Logger) BaseReconciler {
+	return BaseReconciler{
+		client:          client,
+		apiClientReader: apiClientReader,
+		scheme:          scheme,
+		logger:          logger,
+	}
+}
+
+func (b *BaseReconciler) Client() client.Client {
+	return b.client
+}
+
+func (b *BaseReconciler) APIClientReader() client.Reader {
+	return b.apiClientReader
+}
+
+func (b *BaseReconciler) Scheme() *runtime.Scheme {
+	return b.scheme
+}
+
+func (b *BaseReconciler) Logger() logr.Logger {
+	return b.logger
+}
+
+func (b *BaseReconciler) ReconcileResource(desired k8sutils.KubernetesObject, mutateFn MutateFn) error {
+	existing, err := k8sutils.CopyKubernetesObject(desired)
+	if err != nil {
+		return err
+	}
+
+	key, err := client.ObjectKeyFromObject(existing)
+	if err != nil {
+		return err
+	}
+
+	if err = b.Client().Get(context.TODO(), key, existing); err != nil {
```

Comment:
I'm trying to understand the logic of this method.

Why do we copy the `desired` object to the `existing` variable and then we perform a Get on the `existing` variable instead of initializing the `existing` object to an empty object?

Apart from that, in case the object exists  when performing the Get, are we sure no contents are "merged"?

Reply:
`existing` object cannot be initialized to an empty object because you need an object of some specific `Kind`. As I understand, k8s client `Get` method accepts `runtime.Object`, but needs `schema.ObjectKind` set correctly in order to de-serialize properly. Thus, `existing` is created by callling the desired object's `runtime.Object.DeepCopyObject()` that preserves the type of object. 


Regarding your second question. I am assuming that the `existing` object will represent accurately the k8s resource existing in the cluster. Thus, properties will be initialized and  values set from the k8s resource serialization in the response of the request. wdyt?

Reply:
I get the answer to your first question. It makes sense :+1:

Reply:
> Regarding your second question. I am assuming that the existing object will represent accurately the k8s resource existing in the cluster. Thus, properties will be initialized and values set from the k8s resource serialization in the response of the request. wdyt?

I think I understand what you say. You are telling me that the values of existing would be overwritten on top of the  ones that have been defined in the desired. My concern was about the ones that might not be defined too.

To try to give you an example, the situation would be:
* There's a field defined in the desired object but the real existing object does not define it
What would it happen with the contents of that field? would the variable would end up containing the field?

Reply:
I have checked it with the following test:

Create apimanager object:

```
$ k get apimanager apimanager1 -o yaml
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps.3scale.net/v1alpha1","kind":"APIManager","metadata":{"annotations":{},"name":"apimanager1","namespace":"eguzki-metrics"},"spec":{"resourceRequirementsEnabled":false,"wildcardDomain":"eguzki-test.apps.dev-eng-operator-ocp4-2.dev.3sca.net","zync":{"image":"existingImage"}}}
  creationTimestamp: "2020-04-09T10:26:53Z"
  generation: 1
  name: apimanager1
  namespace: eguzki-metrics
  resourceVersion: "19104905"
  selfLink: /apis/apps.3scale.net/v1alpha1/namespaces/eguzki-metrics/apimanagers/apimanager1
  uid: a15c8bc9-7a4c-11ea-b0df-12d443f7f383
spec:
  resourceRequirementsEnabled: false
  wildcardDomain: eguzki-test.apps.dev-eng-operator-ocp4-2.dev.3sca.net
  zync:
    image: existingImage
```

Note the existing fields, `wildcardDomain`, `zync.image`.

Then, I will create a APIManager object:

```
tmpTenant := "mytenantname"                                       
tmpzyncImage := "desiredZyncImage"                                
tmpSystemImage := "desiredSystemImage"                            
desiredAPIManager := &v1alpha1.APIManager{                        
    Spec: v1alpha1.APIManagerSpec{                                
        APIManagerCommonSpec: v1alpha1.APIManagerCommonSpec{      
            TenantName: &tmpTenant,                               
        },                                                        
        Zync: &v1alpha1.ZyncSpec{                                 
            Image: &tmpzyncImage,                                 
        },                                                        
        System: &v1alpha1.SystemSpec{                             
            Image: &tmpSystemImage,                               
        },                                                        
    },                                                            
}                                                                                                                                  
```
Note:
* TenantName new field
* Wildcard property not defined
* Zync.Image property defined and different from existing one
* System.image propoerty new field.

So we have:
* In the existing object:
  * properties not in the desired object: wildcardDomain
  * properties in the desired object: zync.image
* In the desiredObject:
   * properties not in the existing object: tenantName, system.image
   * properties in the existing object: zync.image

The I will run this simple code:

```
fmt.Println("BEFORE")                                                   
s, _ := json.MarshalIndent(desiredAPIManager, "", "\t")                 
fmt.Println(string(s))                                                  
if err := cl.Get(context.TODO(), key, desiredAPIManager); err != nil {  
    panic(err)                                                          
}                                                                       
fmt.Println("AFTER")                                                    
s, _ = json.MarshalIndent(desiredAPIManager, "", "\t")                  
fmt.Println(string(s))                                                  
              
```

And the output is this:

```
BEFORE
{
	"metadata": {
		"creationTimestamp": null
	},
	"spec": {
		"wildcardDomain": "",
		"tenantName": "mytenantname",
		"system": {
			"image": "desiredSystemImage"
		},
		"zync": {
			"image": "desiredZyncImage"
		}
	},
	"status": {
		"deployments": {}
	}
}

AFTER

{
	"metadata": {
		"name": "apimanager1",
		"namespace": "eguzki-metrics",
		"selfLink": "/apis/apps.3scale.net/v1alpha1/namespaces/eguzki-metrics/apimanagers/apimanager1",
		"uid": "a15c8bc9-7a4c-11ea-b0df-12d443f7f383",
		"resourceVersion": "19104905",
		"generation": 1,
		"creationTimestamp": "2020-04-09T10:26:53Z",
		"annotations": {
			"kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"apps.3scale.net/v1alpha1\",\"kind\":\"APIManager\",\"metadata\":{\"annotations\":{},\"name\":\"apimanager1\",\"namespace\":\"eguzki-metrics\"},\"spec\":{\"resourceRequirementsEnabled\":false,\"wildcardDomain\":\"eguzki-test.apps.dev-eng-operator-ocp4-2.dev.3sca.net\",\"zync\":{\"image\":\"existingImage\"}}}\n"
		}
	},
	"spec": {
		"wildcardDomain": "eguzki-test.apps.dev-eng-operator-ocp4-2.dev.3sca.net",
		"tenantName": "mytenantname",
		"resourceRequirementsEnabled": false,
		"system": {
			"image": "desiredSystemImage"
		},
		"zync": {
			"image": "existingImage"
		}
	},
	"status": {
		"deployments": {}
	}
}
```

Conclusion:
* Properties in the existing object and not in the desired object, wilcarddomain: updated correctly in desired object [OK]
* Properties not in the existing object and in the desired object, tenant name and system.image: Exist in the desired object after `Get` call. [NOT OK]
* Properties in the existing object and in the desired object: zync.image: updated correctly in desired object [OK]

So, I guess this is a problem because attributes in the desired object not existing in the cluster will still be there after reading from the cluster. I will change the signature of the function from
```
func (b *BaseReconciler) ReconcileResource(desired k8sutils.KubernetesObject, mutateFn MutateFn) error
```

to 

```
func (b *BaseReconciler) ReconcileResource(obj, desired k8sutils.KubernetesObject, mutateFn MutateFn) error
```
And it is responsability of the callee to pass "empty" object with a correct Kind.

Something like this:

```
obj := &appsv1.Deployment{}
err = r.ReconcileResource(obj, desiredDeployment, DeploymentMutator)
```

Reply:
:+1: 

Thank you for looking into this and trying it.

That seems to be the solution yes. It's not as clean but it's how k8s implements deserialization in the receiving parameter in the Get so not much we can do about that I think unfortunately.

Reply:
implemented that way

Reply:
:+1: